lib/protobuf/verifier.ex

```diff
@@ -0,0 +1,265 @@
+  defp repeated_or_not(val, false = _repeated, func), do: [func.(val)]
+
+  # NOTE: This can aggregate all of the errors if we want
+  defp first_non_ok_value_if_present([]), do: :ok
+  defp first_non_ok_value_if_present([:ok | rest]), do: first_non_ok_value_if_present(rest)
+  defp first_non_ok_value_if_present([non_ok_value | _rest]), do: non_ok_value
+
+  @spec class_field(map) :: atom
+  defp class_field(%{wire_type: wire_delimited(), embedded?: true}), do: :embedded
+  defp class_field(_), do: :normal
+
+  @spec verify_type(atom, any) :: :ok
+  def verify_type(:string, n) when is_binary(n), do: :ok
+  def verify_type(:bool, true), do: :ok
+  def verify_type(:bool, false), do: :ok
+  def verify_type(:float, :infinity), do: :ok
+  def verify_type(:float, :negative_infinity), do: :ok
+  def verify_type(:float, :nan), do: :ok
+  def verify_type(:float, n) when is_number(n), do: :ok
+  def verify_type(:double, :infinity), do: :ok
+  def verify_type(:double, :negative_infinity), do: :ok
+  def verify_type(:double, :nan), do: :ok
+  def verify_type(:double, n) when is_number(n), do: :ok
+  def verify_type(:bytes, n) when is_binary(n), do: :ok
+  def verify_type(:int32, n) when is_integer(n) and n >= -0x80000000 and n <= 0x7FFFFFFF, do: :ok
+
+  def verify_type(:int64, n)
+      when is_integer(n) and n >= -0x8000000000000000 and n <= 0x7FFFFFFFFFFFFFFF,
+      do: :ok
+
+  def verify_type(:uint32, n) when is_integer(n) and n >= 0 and n <= 0xFFFFFFFF, do: :ok
+  def verify_type(:uint64, n) when is_integer(n) and n >= 0 and n <= 0xFFFFFFFFFFFFFFFF, do: :ok
+  def verify_type(:sint32, n) when is_integer(n) and n >= -0x80000000 and n <= 0x7FFFFFFF, do: :ok
+
+  def verify_type(:sint64, n)
+      when is_integer(n) and n >= -0x8000000000000000 and n <= 0x7FFFFFFFFFFFFFFF,
+      do: :ok
+
+  def verify_type(:fixed64, n) when is_integer(n) and n >= 0 and n <= 0xFFFFFFFFFFFFFFFF, do: :ok
+
+  def verify_type(:sfixed64, n)
+      when is_integer(n) and n >= -0x8000000000000000 and n <= 0x7FFFFFFFFFFFFFFF,
+      do: :ok
+
+  def verify_type(:fixed32, n) when is_integer(n) and n >= 0 and n <= 0xFFFFFFFF, do: :ok
+
+  def verify_type(:sfixed32, n) when is_integer(n) and n >= -0x80000000 and n <= 0x7FFFFFFF,
+    do: :ok
+
+  def verify_type({:enum, type}, n) when is_atom(n) do
+    if type.mapping() |> Map.has_key?(n) do
+      :ok
+    else
+      {:error, "#{inspect(n)} is not a valid value in enum #{type}"}
+    end
+  end
+
+  def verify_type({:enum, type}, n) when is_integer(n) do
+    if type.__reverse_mapping__() |> Map.has_key?(n) do
+      :ok
+    else
+      {:error, "#{inspect(n)} is not a valid value in enum #{type}"}
+    end
+  end
+
+  # Enum failure case
+  def verify_type({:enum, type}, n) do
+    {:error, "#{inspect(n)} is invalid for type #{type}"}
+  end
+
+  # General failure case
+  def verify_type(type, n) do
+    {:error, "#{inspect(n)} is invalid for type #{type}"}
```

Comment:
Although this would be very helpful for debugging, I wonder if there'd be the possibility of accidental PII leakage with this

Reply:
I agree, maybe just say the value doesnt match type.

Reply:
Sad to see it go but I took the inspects out